[READY] Overhaul Go completer

micbou opened the pull request:
This PR includes the changes mentioned in PR #282 for the Go completer and other small changes (listed in the commit message). The main change is the way we start the `gocode` daemon: instead of letting the client automatically start the daemon, we directly start it by running the `gocode` executable in server mode (using the `-s` parameter). This allows us to:
 - get its PID;
 - log stdout and stderr into files;
 - enable `gocode` debug mode;
 - set the port listened by `gocode`;
 - terminate the process;
 - implement the `_ServerIs*` methods;
 - etc.

The second important change is the implementation of the `DebugInfo` method to show debug informations about the Go completer. The way informations are displayed will be adapted to other completers. Here's the output obtained in Vim when editing a Go file before and after the changes:

Before:
```
Printing YouCompleteMe debug information...
-- Server has Clang support compiled in: True
-- Clang version: clang version 3.8.0 (branches/release_38)
--
-- Server running at: http://127.0.0.1:59642
-- Server process ID: 4884
-- Server logfiles:
--   c:\users\micbou\appdata\local\temp\ycm_temp\server_59642_stdout.log
--   c:\users\micbou\appdata\local\temp\ycm_temp\server_59642_stderr.log
```
*How can I write a good issue report in these conditions?*

After:
```
Printing YouCompleteMe debug information...
-- Server has Clang support compiled in: True
-- Clang version: clang version 3.8.0 (branches/release_38)
-- Go completer debug information:
--   Gocode running at: http://127.0.0.1:59667
--   Gocode process ID: 8720
--   Gocode binary: C:\Users\micbou\VIM~1\bundle\YOUCOM~2\THIRD_~1\ycmd\third_pa
rty\gocode\gocode.exe
--   Gocode logfiles:
--     C:\Users\micbou\AppData\Local\Temp\ycm_temp\gocode_59667_stdout.log
--     C:\Users\micbou\AppData\Local\Temp\ycm_temp\gocode_59667_stderr.log
--   Godef binary: C:\Users\micbou\VIM~1\bundle\YOUCOM~2\THIRD_~1\ycmd\third_par
ty\godef\godef.exe
-- Server running at: http://127.0.0.1:59663
-- Server process ID: 2788
-- Server logfiles:
--   c:\users\micbou\appdata\local\temp\ycm_temp\server_59663_stdout.log
--   c:\users\micbou\appdata\local\temp\ycm_temp\server_59663_stderr.log
```
*Please help, I am overwhelmed by all this information.*

Finally, the last important change (copied from the Python completer) is replacing the `StartServer` subcommand with the `RestartServer` one (reasoning is that if you can restart the server, you can start it too) and hiding the `StopServer` subcommand from the user (if users need to stop the server then something is wrong with the completer).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/505)
<!-- Reviewable:end -->

Valloric commented:
Excellent stuff! Thanks for taking this on.

Some comments left.

<details><summary><i>Previously, micbou wrote…</i></summary>
> #### \[READY\] Overhaul Go completer
> 
> This PR includes the changes mentioned in PR #282 for the Go completer and other small changes (listed in the commit message). The main change is the way we start the `gocode` daemon: instead of letting the client automatically start the daemon, we directly start it by running the `gocode` executable in server mode (using the `-s` parameter). This allows us to:
> 
>  - get its PID;
> 
>  - log stdout and stderr into files;
> 
>  - enable `gocode` debug mode;
> 
>  - set the port listened by `gocode`;
> 
>  - terminate the process;
> 
>  - implement the `_ServerIs*` methods;
> 
>  - etc.
> 
> 
> 
> The second important change is the implementation of the `DebugInfo` method to show debug informations about the Go completer. The way informations are displayed will be adapted to other completers. Here's the output obtained in Vim when editing a Go file before and after the changes:
> 
> 
> 
> Before:
> 
> ```
> 
> Printing YouCompleteMe debug information...
> 
> -- Server has Clang support compiled in: True
> 
> -- Clang version: clang version 3.8.0 (branches/release_38)
> 
> --
> 
> -- Server running at: http://127.0.0.1:59642
> 
> -- Server process ID: 4884
> 
> -- Server logfiles:
> 
> --   c:\users\micbou\appdata\local\temp\ycm_temp\server_59642_stdout.log
> 
> --   c:\users\micbou\appdata\local\temp\ycm_temp\server_59642_stderr.log
> 
> ```
> 
> *How can I write a good issue report in these conditions?*
> 
> 
> 
> After:
> 
> ```
> 
> Printing YouCompleteMe debug information...
> 
> -- Server has Clang support compiled in: True
> 
> -- Clang version: clang version 3.8.0 (branches/release_38)
> 
> -- Go completer debug information:
> 
> --   Gocode running at: http://127.0.0.1:59667
> 
> --   Gocode process ID: 8720
> 
> --   Gocode binary: C:\Users\micbou\VIM~1\bundle\YOUCOM~2\THIRD_~1\ycmd\third_pa
> 
> rty\gocode\gocode.exe
> 
> --   Gocode logfiles:
> 
> --     C:\Users\micbou\AppData\Local\Temp\ycm_temp\gocode_59667_stdout.log
> 
> --     C:\Users\micbou\AppData\Local\Temp\ycm_temp\gocode_59667_stderr.log
> 
> --   Godef binary: C:\Users\micbou\VIM~1\bundle\YOUCOM~2\THIRD_~1\ycmd\third_par
> 
> ty\godef\godef.exe
> 
> -- Server running at: http://127.0.0.1:59663
> 
> -- Server process ID: 2788
> 
> -- Server logfiles:
> 
> --   c:\users\micbou\appdata\local\temp\ycm_temp\server_59663_stdout.log
> 
> --   c:\users\micbou\appdata\local\temp\ycm_temp\server_59663_stderr.log
> 
> ```
> 
> *Please help, I am overwhelmed by all this information.*
> 
> 
> 
> Finally, the last important change (copied from the Python completer) is replacing the `StartServer` subcommand with the `RestartServer` one (reasoning is that if you can restart the server, you can start it too) and hiding the `StopServer` subcommand from the user (if users need to stop the server then something is wrong with the completer).
> 
> 
> 
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[ycmd/completers/go/go_completer.py, line 127 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOI1GVp8SIlBZax3SW:-KIOI1GVp8SIlBZax3SX:-1032703792) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L127)):*
> ```Python
> 
> 
>   def _ComputeOffset( self, contents, line, col ):
> ```

This does'n seem like it needs to be a member func; it can just be a private func in the module.

The benefit if not making it a member func is that it's obvious it doesn't mutate any GoCompleter object state, improving readability & maintainability.

---

*[ycmd/completers/go/go_completer.py, line 130 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOHeoY3iBkAI9UDdUM:-KIOHeoY3iBkAI9UDdUN:-594000511) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L130)):*
> ```Python
>     """Compute the byte offset in the file given the line and column."""
>     contents = ToBytes( contents )
>     curline = 1
> ```

I'd expand the name to `current_line`. Similar for line below. 

---

*[ycmd/completers/go/go_completer.py, line 142 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOHrLEhYuEyFdg8puo:-KIOHrLEhYuEyFdg8pup:-2109397076) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L142)):*
> ```Python
>     _logger.error( 'Go completer could not compute byte offset '
>                    'corresponding to L%i C%i', line, col )
>     return -1
> ```

Is returning -1 ok? No need for an exception to abort the request?

---

*[ycmd/completers/go/go_completer.py, line 145 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIFYjmggu-soij5y8:-KIOIFYjmggu-soij5y9:1935797553) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L145)):*
> ```Python
> 
> 
>   def _ConvertCompletionData( self, completion_data ):
> ```

Similar for _ComputeOffset; doesn't seem like it needs to be a member func.

---

*[ycmd/completers/go/go_completer.py, line 316 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIpdYvUorEFvavO0x:-KIOIpdYvUorEFvavO0y:-162499588) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L316)):*
> ```Python
>                                        '-o={0}'.format( offset ) ],
>                                      contents = contents )
>     except RuntimeError as error:
> ```

Can we catch a more specific exception type if possible?

---

*[ycmd/completers/go/go_completer.py, line 351 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIzQMzuc8NbpiYF6Y:-KIOIzQMzuc8NbpiYF6Z:-1103816104) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L351)):*
> ```Python
>                               'status' ] )
>       return True
>     except RuntimeError as error:
> ```

More specific exception?

---

*[ycmd/completers/go/go_completer.py, line 370 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOJ456HA6qPa-y9Mqt:-KIOJ456HA6qPa-y9Mqu:-112554355) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L370)):*
> ```Python
>                '    {3}\n'
>                '    {4}\n'
>                '  Godef binary: {5}'.format( self._gocode_address,
> ```

Probably needs to be wrapped in the lock since it access the data protected by it. Both reads & writes need the lock.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KIOJKVtJMcs2aPU4e64:1008456488)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 5 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[ycmd/completers/go/go_completer.py, line 127 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOI1GVp8SIlBZax3SW:-KIOO6gvY47qbBnl-6XG:-247167634) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L127)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> This does'n seem like it needs to be a member func; it can just be a private func in the module.
> 
> The benefit if not making it a member func is that it's obvious it doesn't mutate any GoCompleter object state, improving readability & maintainability.
</details>
I reverted the change.

---

*[ycmd/completers/go/go_completer.py, line 130 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOHeoY3iBkAI9UDdUM:-KIOODOk9vZdxIxK_3MA:495613066) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L130)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> I'd expand the name to `current_line`. Similar for line below. 
</details>
Done. For `curcol` too.

---

*[ycmd/completers/go/go_completer.py, line 142 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOHrLEhYuEyFdg8puo:-KIOOMB5FWolIhVM6wkx:1556488574) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L142)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Is returning -1 ok? No need for an exception to abort the request?
</details>
I replaced `offset` to `-1` to check `gocode` behavior and it panicked so, as you suggested, raising an exception here is better than catching the one from `gocode`.

---

*[ycmd/completers/go/go_completer.py, line 145 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIFYjmggu-soij5y8:-KIOPCcyTyCTbsUbV8G6:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L145)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Similar for _ComputeOffset; doesn't seem like it needs to be a member func.
</details>
Done.

---

*[ycmd/completers/go/go_completer.py, line 316 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIpdYvUorEFvavO0x:-KIOPEKYFCJmCh5VZG32:1424158230) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L316)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Can we catch a more specific exception type if possible?
</details>
No because we want to catch the `RuntimeError` exception that may be raised in the `_ExecuteCommand` method. Also, we are logging the exception so we should have all the information we need to debug the exception.

---

*[ycmd/completers/go/go_completer.py, line 351 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIzQMzuc8NbpiYF6Y:-KIOPN2AQsbxwClM_E-E:-647164044) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L351)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> More specific exception?
</details>
See my comment above.

---

*[ycmd/completers/go/go_completer.py, line 370 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOJ456HA6qPa-y9Mqt:-KIOQR4TkXcoLeagxMST:318107366) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L370)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Probably needs to be wrapped in the lock since it access the data protected by it. Both reads & writes need the lock.
</details>
We'll need to do the same for other completers using a lock.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/completers/go/go_completer.py, line 316 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIpdYvUorEFvavO0x:-KITm5hvFamAZfUre3i9:-1836429261) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L316)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> No because we want to catch the `RuntimeError` exception that may be raised in the `_ExecuteCommand` method. Also, we are logging the exception so we should have all the information we need to debug the exception.
</details>
Ok, but leave a comment about it.

---

*[ycmd/completers/go/go_completer.py, line 351 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIzQMzuc8NbpiYF6Y:-KITmDPPqFKcjx1X3i91:-184073581) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L351)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> See my comment above.
</details>
Comment needed.

---

*[ycmd/completers/go/go_completer.py, line 370 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOJ456HA6qPa-y9Mqt:-KITmWfotx4aAi4W8lyY:-1530771663) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L370)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We'll need to do the same for other completers using a lock.
</details>
Agreed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6282151/badge)](https://coveralls.io/builds/6282151)
> 
> Coverage decreased (-0.08%) to 84.878% when pulling **f401a2676a814d57ff14d426aca2a6229bcc491a on micbou:go-completer-overhaul** into **994b7ebcf8bb105734259557d134dc78addfbc05 on Valloric:master**.
> 
</details>


---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KIUjlhSeLwLOHppE5j5:980077423)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/completers/go/go_completer.py, line 316 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIpdYvUorEFvavO0x:-KIYf1egR-gb7fM0Wbnb:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L316)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Ok, but leave a comment about it.
</details>
Done.

---

*[ycmd/completers/go/go_completer.py, line 351 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIzQMzuc8NbpiYF6Y:-KIYf2AwUfkYnJ1YDgek:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L351)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Comment needed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I see coverage decreasing. Not good. Could we add some more tests? Idea is to aim for >=95% test coverage for new code.

micbou commented:
I added tests for the `DebugInfo` method, a test for `_ComputeOffset` when the offset cannot be computed, and a test for when `GoTo` cannot jump to definition.

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6320363/badge)](https://coveralls.io/builds/6320363)
> 
> Coverage increased (+0.3%) to 85.246% when pulling **5ff026cc847b870e8ffa5e4868873546ecfe0dc5 on micbou:go-completer-overhaul** into **22f229c91f7e3e6b60da521104bc0a31fc72951d on Valloric:master**.
> 
</details>


---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KIbnF5Mdz8VXpgvAkNU:1454181712)*
<!-- Sent from Reviewable.io -->

micbou commented:
Move the `DebugInfo` tests to a separate file. They are unrelated to subcommands.

<details><summary><i>Previously, micbou wrote…</i></summary>
> I added tests for the `DebugInfo` method, a test for `_ComputeOffset` when the offset cannot be computed, and a test for when `GoTo` cannot jump to definition.
</details>


---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KIdv8cVDPThTIIStP8v:1460691641)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6336036/badge)](https://coveralls.io/builds/6336036)
> 
> Coverage increased (+0.3%) to 85.246% when pulling **dd3463b713d4c9166610240d399ace38bd6c4bee on micbou:go-completer-overhaul** into **22f229c91f7e3e6b60da521104bc0a31fc72951d on Valloric:master**.
> 
</details>


---

Reviewed 3 of 5 files at r1, 1 of 1 files at r3, 1 of 3 files at r4, 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/completers/go/go_completer.py, line 316 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIpdYvUorEFvavO0x:-KIe1qJ24NG8m_zsdYtc:-1460763861) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L316)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Done.
</details>
I don't think I understood the comment :confused: can we reword it a little? :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KIe3Yo1wTzfxAbi6myw:1417088721)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/completers/go/go_completer.py, line 316 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/505#-KIOIpdYvUorEFvavO0x:-KIeCgLNXbcZgkoLFamU:888953220) ([raw file](https://github.com/valloric/ycmd/blob/daed0ff5cf1830cd6461e1b12cee4bdcdf0b11f9/ycmd/completers/go/go_completer.py#L316)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I don't think I understood the comment :confused: can we reword it a little? :(
</details>
I reworded it. Hope it is more understandable this way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6337191/badge)](https://coveralls.io/builds/6337191)
> 
> Coverage increased (+0.3%) to 85.246% when pulling **9bb274546421070eff45c01b34a89065b37363b6 on micbou:go-completer-overhaul** into **22f229c91f7e3e6b60da521104bc0a31fc72951d on Valloric:master**.
> 
</details>


---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KIibuqIltcz_w7y6qYh:1417088721)*
<!-- Sent from Reviewable.io -->

Valloric commented:
@micbou Any specific reason why you're not merging this? You have two LGTMs.

micbou commented:
No particular reason. Let's merge it.

@homu r+

<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> @micbou Any specific reason why you're not merging this? You have two LGTMs.
</details>


---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/505#-:-KJbweCuAyVHO9Z1cY7e:-1765557051)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 3fe7ea5 has been approved by `micbou`

<!-- @homu r=micbou 3fe7ea5f60bba1c27c06ab6d239c4184cd8ee5d9 -->

homu commented:
:hourglass: Testing commit 3fe7ea5 with merge 24ccb3b...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/135732240)
